src/perf/lib/Tcp.cpp

```diff
@@ -0,0 +1,724 @@
+{
+    TcpWorker* This = (TcpWorker*)Context;
+
+    while (!This->Engine->Shutdown) {
+        TcpConnection* Connection;
+        CxPlatDispatchLockAcquire(&This->Lock);
+        if (CxPlatListIsEmpty(&This->Connections)) {
+            Connection = nullptr;
+        } else {
+            Connection =
+                CXPLAT_CONTAINING_RECORD(
+                    CxPlatListRemoveHead(&This->Connections),
+                    TcpConnection,
+                    Entry);
+            Connection->Entry.Flink = nullptr;
+        }
+        CxPlatDispatchLockRelease(&This->Lock);
+        if (Connection) {
+            Connection->Process();
+            Connection->Release();
+        } else {
+            CxPlatEventWaitForever(This->WakeEvent);
+        }
+    }
+
+    CXPLAT_THREAD_RETURN(0);
+}
+
+void TcpWorker::QueueConnection(TcpConnection* Connection)
+{
+    CxPlatDispatchLockAcquire(&Lock);
+    if (!Connection->Entry.Flink) {
+        Connection->AddRef();
+        CxPlatListInsertTail(&Connections, &Connection->Entry);
+        CxPlatEventSet(WakeEvent);
+    }
+    CxPlatDispatchLockRelease(&Lock);
+}
+
+// ############################# SERVER #############################
+
+TcpServer::TcpServer(TcpEngine* Engine, const QUIC_CREDENTIAL_CONFIG* CredConfig) :
+    Initialized(false), Engine(Engine), SecConfig(nullptr), Listener(nullptr)
+{
+    LoadSecConfigHelper Helper;
+    if ((SecConfig = Helper.Load(CredConfig)) == nullptr) {
+        return;
+    }
+    if (QUIC_FAILED(
+        CxPlatSocketCreateTcpListener(
+            Engine->Datapath,
+            nullptr,
+            this,
+            &Listener))) {
+        return;
+    }
+    Initialized = true;
+}
+
+TcpServer::~TcpServer()
+{
+    if (Listener) {
+        CxPlatSocketDelete(Listener);
+    }
+    if (SecConfig) {
+        CxPlatTlsSecConfigDelete(SecConfig); // TODO - Ref counted instead?
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Function_class_(CXPLAT_DATAPATH_ACCEPT_CALLBACK)
+void
+TcpServer::AcceptCallback(
+    _In_ CXPLAT_SOCKET* /* ListenerSocket */,
+    _In_ void* ListenerContext,
+    _In_ CXPLAT_SOCKET* AcceptSocket,
+    _Out_ void** AcceptClientContext
+    )
+{
+    TcpServer* This = (TcpServer*)ListenerContext;
+    *AcceptClientContext = new TcpConnection(This->Engine, This->SecConfig, AcceptSocket);
+}
+
+// ############################ CONNECTION ############################
+
+TcpConnection::TcpConnection(
+    TcpEngine* Engine,
+    const QUIC_CREDENTIAL_CONFIG* CredConfig,
+    const QUIC_ADDR* RemoteAddress,
+    const QUIC_ADDR* LocalAddress,
+    void* Context) :
+    Server(false), Initialized(false), IndicateAccept(false), IndicateConnect(false),
+    IndicateDisconnect(false), StartTls(false),
+    Engine(Engine), Worker(nullptr), Socket(nullptr), SecConfig(nullptr), Tls(nullptr),
+    ReceiveData(nullptr), SendData(nullptr), BatchedSendData(nullptr),
+    BufferedDataLength(0), Context(Context)
+{
+    Entry.Flink = NULL;
```

Comment:
Why is this not using CxPlatListInitializeHead? It looks like CxPlatListInsertTail would assert in this case, because Blink would be garbage, and fail QuicListEntryValidate (Also I need to fix QuicListEntryValidate)

Reply:
If this doesn't need to be circular, why not use CXPLAT_SLIST_ENTRY?

Reply:
Wait nevermind I'm misunderstanding what this does. This is the entry, not the list.